### PR TITLE
Fixed #13251 -- Made pre/post_delete signals dispatch the origin.

### PR DESCRIPTION
**Pull request by mgaligniana**
This is an old ticket! [13251](https://code.djangoproject.com/ticket/13251)

But here I add an approach (if it still needed)

### CODE REVIEW

File: django/db/models/base.py [hunk 1]
```diff
@@ -987,7 +987,7 @@ def delete(self, using=None, keep_parents=False):
                 "to None." % (self._meta.object_name, self._meta.pk.attname)
             )
         using = using or router.db_for_write(self.__class__, instance=self)
-        collector = Collector(using=using)
+        collector = Collector(using=using, origin=self.__class__)
```
Comment (felixxm) [inline django/db/models/base.py:990]: :thinking: 
```suggestion
        collector = Collector(using=using, origin=self)
```

File: docs/ref/signals.txt [hunk 2]
```diff
@@ -219,6 +222,9 @@ Arguments sent with this signal:
 ``using``
     The database alias being used.
 
+``origin``
+    Model or QuerySet class depending where it was sent from
```
Comment (felixxm) [inline docs/ref/signals.txt:226]: `versionadded` annotation is missing.

File: docs/ref/signals.txt [hunk 3]
```diff
@@ -195,6 +195,9 @@ Arguments sent with this signal:
...
```
Comment (felixxm) [inline docs/ref/signals.txt:199]: `versionadded` annotation is missing.

**Review by felixxm — COMMENTED** [review]
@mgaligniana Thanks for this patch :+1:

As far as I'm aware it was crucial for the report to get an instance of a `Model` or `QuerySet`, not only a class (see [proposed patch](https://code.djangoproject.com/attachment/ticket/13251/changes.patch)). I don't see much value in getting `QuerySet` in the `origin`. Please also add release notes.

File: tests/signals/tests.py [hunk 4]
```diff
@@ -167,6 +167,37 @@ def __call__(self, signal, sender, instance, **kwargs):
...
+            Person.objects.create(first_name="John", last_name="Smith")
+            Person.objects.all().delete()
+            self.assertEqual(data, [
+                models.QuerySet,
+                models.QuerySet,
+            ])
+
+        finally:
+            signals.pre_delete.disconnect(pre_delete_handler)
+            signals.post_delete.disconnect(post_delete_handler)
```
[the remark this answers is not in the document]
Comment (mgaligniana) [reply tests/signals/tests.py:238]: Added! Let me know if they are correct!

**Review by felixxm — COMMENTED** [review]
@mgaligniana Thanks :+1: 

I added `origin` to `Collector` subclasses, and pushed small edits to tests and docs.

File: tests/signals/tests.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (felixxm) [reply tests/signals/tests.py:240]: Happy New Year :fireworks:

**Pull request merged**